File: BingItOn.java
import java.io.*;

public class BingItOn {
  public static void main(String[] args) throws IOException{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int num = Integer.parseInt(br.readLine()) ;
    Node trie = new Node();

    for(int i = 0; i<num; i++){
      String s = br.readLine();

      trie.add(s,0);
      //System.out.println(trie.findCount(s,0) - 1);
      Integer temp = trie.findCount(s,0) - 1;
      bw.write(temp.toString() +"\n");
    }
    bw.flush(); 
  }

  public static class Node {
    private int numOfChars = 26;
    Node children[] = new Node[numOfChars];
    int size = 0;

    public Node getNode(char c){
      return children[getCharIndex(c)];
    }

    public void setNode(char c, Node n){
      children[getCharIndex(c)] = n;
    }

    public void add(String s, int index){
       size++;
       if(index == s.length()) return;

       char current = s.charAt(index);
       int charCode = getCharIndex(current);
       Node child = getNode(current);

       if(child==null){
         child = new Node();
         setNode(current, child);
       }

       child.add(s, index+1);
    }

    public int findCount(String s, int index){
      if(index == s.length()) return size;

      Node child = getNode(s.charAt(index));

      if(child == null){
        return 0;
      }

      return child.findCount(s, index + 1);
    }

    public int getCharIndex(char c){
      return (c - 'a');
    }
  }

}

File: BingItOn.cpp
#include <bits/stdc++.h>

using namespace std; 

struct node{
  node * children[26];
  
  int count; 
  
  node(){
    count = 0;
    for(int i = 0; i<26; i++)
      children[i] = NULL; 
  }
};


void add( node* n, char* s, int index){
  
    if(*s){
    int i = *s - 'a'; 
    
    if(n->children[i] == NULL){
      n->children[i] = new node();
    }
    
    n->children[i]->count++; 
    return add(n->children[i], s+1, index+1); 
    }
    
    
      return; 
}


int findCount( node * n, char* s, int index){

    if(*s){
  int i = (*s - 'a');
  
  return findCount(n->children[i], s+1, index+1); 
    }
    
    
    return n->count; 
}

int main() {
   node* n = new node(); 
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

  int t;
  cin >> t; 

  for(int i = 0; i<t; i++){
    char * line = new char[32]; 

    cin >> line; 

    add(n, line, 0);

    printf("%d\n", findCount(n,line,0)-1);
}

  return 0; 
}
